caffe2/ideep/operators/quantization/int8_conv_op.cc

```diff
@@ -96,18 +95,55 @@ class IDEEPInt8ConvOp : public IDEEPConvPoolOpBase {
       }
     }
 
-    if (InputSize() > last_input_) {
-      ideep::convolution_forward::compute(
-          op_key_, X, filter_, bias_, Y_dims, *Y,
-          stride_, dilation_, pad_tl(), pad_br(), group_,
-          iscale(), iscale(), Y_scales_, attr_, algo_,
-          iprop::forward_inference, ipadding::zero, lowp_kind_);
+    bool with_bias = InputSize() > last_input_;
+    if (input_changed || weights_changed) {
+      auto Y_dims = CalcOutputDims(X, filter.get_dim(0));
+      if (with_bias) {
+        ideep::convolution_forward::prepare(
```

Comment:
What does `ideep::convolution_forward::prepare` do?

Reply:
Now `ideep::conv` has two kind of API:

1. Single-call functional API (exactly the same as before):

  `conv::compute(src, weight, dst, group, stride, padding, algorithm, scale, ...)`

  It's been used in pytorch now. The downside is it didn't carry any state info that can be used for subsequent calls to it. If we'd like to **cache conv `pd` and pre-computed requantization scales**, there's no way to pass them.

Implementation wise, it's basically a combination of the following 2-staged api calls.

2. Two-staged API. 

`conv::prepare(conv_param, src, weight, dst, group, stride, padding, algorithm, scale, ...)`
`conv::compute(conv_param, src, weight, dst)`

It has an additional param called `conv_param` that's handy for carrying any info we need. Otherwise, there would be a huge gap in caffe2 models.

Sidenote:
The previous API is a little bit confusing from the fwk perspective
```
conv::compute(key, src, weight, dst, group, stride, padding, algorithm, scale, ...)
```
The actual internal control flow depends on the `key`:
- if the `key` is empty, it will also do the 2-stage `prepare` and `compute` internally
- if the `key` has value, it will dispatch to the `compute` only

Reply:
well noted, thanks!